src/Umbraco.Infrastructure/Runtime/FileSystemMainDomLock.cs

```diff
@@ -0,0 +1,147 @@
+                    return Task.FromResult(true);
+                }
+                catch (IOException)
+                {
+                    _logger.LogDebug("Couldn't obtain MainDom lock file handle, signalling for release of {lockFilePath}", _lockFilePath);
+                    CreateLockReleaseFile();
+                    Thread.Sleep(500);
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogError(ex, "Unexpected exception attempting to obtain MainDom lock file handle {lockFilePath}, giving up", _lockFilePath);
+                    return Task.FromResult(false);
+                }
+            }
+            while (stopwatch.ElapsedMilliseconds < millisecondsTimeout);
+
+            return Task.FromResult(false);
+        }
+
+        // Create a long running task to poll to check if anyone has created a lock release file.
+        public Task ListenAsync()
+        {
+            if (_listenForReleaseSignalFileTask != null)
+            {
+                return _listenForReleaseSignalFileTask;
+            }
+
+            _listenForReleaseSignalFileTask = Task.Factory.StartNew(
+                ListeningLoop,
+                _cancellationTokenSource.Token,
+                TaskCreationOptions.LongRunning,
+                TaskScheduler.Default);
+
+            return _listenForReleaseSignalFileTask;
+        }
+
+        public void Dispose()
+        {
+            _lockFileStream?.Close();
+            _lockFileStream = null;
+        }
+
+        private void CreateLockReleaseFile()
+        {
+            try
+            {
+                // Dispose immediately to release the file handle so it's easier to cleanup in any process.
+                using FileStream releaseFileStream = File.Open(_releaseSignalFilePath, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.ReadWrite);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Unexpected exception attempting to create lock release signal file {file}", _releaseSignalFilePath);
+            }
+        }
+
+        private void DeleteLockReleaseFile()
+        {
+            List<Exception> encounteredExceptions = new();
+            for (var i = 0; i < s_maxTriesRemovingLockReleaseSignalFile; i++)
+            {
+                try
+                {
+                    File.Delete(_releaseSignalFilePath);
+                    return;
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogError(ex, "Unexpected exception attempting to delete release signal file {file}", _releaseSignalFilePath);
+                    encounteredExceptions.Add(ex);
+                    Thread.Sleep(500 * (i + 1));
+                }
+            }
+
+            throw new ApplicationException($"Failed to remove lock release signal file {_releaseSignalFilePath}", new AggregateException(encounteredExceptions));
```

Comment:
And this keeps retrying the operation, even when the timeout passed to `AcquireLockAsync` has long expired. Hence my previous comment about using a `CancellationTokenSource` with a timeout and passing the token into these methods, so it can exit early.